### ISSUE TITLE
Error loading simple LUA script in EMQ 3

### ISSUE THREAD

**Issue by jdavidagudelo** [emqx/emqx#1940]
#### Environment

- OS: Ubuntu 16.04
- Erlang/OTP: 21
- EMQ: 3.0-rc2

#### Description

Error loading simple LUA script.

I was trying to load the following LUA script to EMQ 3:

```
function on_message_publish(clientid, username, topic, payload, qos,retained)
    return topic, payload, qos, true
end

function register_hook()
    return "on_message_publish"
end
```

I copy the script to hook_lua and then I try to run:

> ./bin/emqx_ctl plugins load emqx_lua_hook

Then the next error appears:

```
load plugin error: {emqx_lua_hook,
                    {bad_return,
                     {{emqx_lua_hook_app,start,[normal,[]]},
                      {'EXIT',
                       {{function_clause,
                         [{emqx_hooks,add,
                           ['message.publish',
                            {"hook_lua/update_publish_topic.lua",
                             fun emqx_lua_script:on_message_publish/2},
                            [{luerl,
                              {array,13,100,undefined,
                               {{{table,
                                  {array,0,10,nil,10},
                                  {{{{empty,<<"_G">>,
                                      {tref,0},
                                      empty,<<"_VERSION">>,<<"Lua 5.2">>,
                                      empty},
                                     <<"assert">>,
                                     {function,
                                      #Fun<luerl_lib_basic.1.127600018>},
                                     {empty,<<"bit32">>,{tref,5},empty},
                                     <<"collectgarbage">>,
                                     {function,
                                      #Fun<luerl_lib_basic.2.127600018>},
                                     {empty,<<"debug">>,{tref,12},empty}},
                                    <<"dofile">>,
                                    {function,
                                     #Fun<luerl_lib_basic.3.127600018>},
                                    {{empty,<<"eprint">>,
                                      {function,
                                       #Fun<luerl_lib_basic.4.127600018>},
                                      empty},
                                     <<"error">>,
                                     {function,
                                      #Fun<luerl_lib_basic.0.127600018>},
                                     {empty,<<"getmetatable">>,
                                      {function,
                                       #Fun<luerl_lib_basic.5.127600018>},
                                      empty,<<"io">>,
                                      {tref,6},
                                      empty}}},
                                   <<"ipairs">>,
                                   {function,
                                    #Fun<luerl_lib_basic.6.127600018>},
                                   {{{empty,<<"load">>,
                                      {function,
                                       #Fun<luerl_lib_basic.7.127600018>},
                                      empty},
                                     <<"loadfile">>,
                                     {function,
                                      #Fun<luerl_lib_basic.8.127600018>},
                                     {empty,<<"loadstring">>,
                                      {function,
                                       #Fun<luerl_lib_basic.9.127600018>},
                                      empty,<<"math">>,
                                      {tref,7},
                                      empty}},
                                    <<"next">>,
                                    {function,
                                     #Fun<luerl_lib_basic.10.127600018>},
                                    {{empty,<<"on_message_publish">>,
                                      {function,6,0,[],
                                       [1,2,3,4,5,6],
                                       [{push_lvar,1,3},
                                        {push_lvar,1,4},
                                        {push_lvar,1,5},
                                        {push_last_lit,true},
                                        {return,4}]},
                                      empty,<<"os">>,
                                      {tref,8},
                                      empty},
                                     <<"package">>,
                                     {tref,4},
                                     {empty,<<"pairs">>,
                                      {function,
                                       #Fun<luerl_lib_basic.11.127600018>},
                                      empty},
                                     <<"pcall">>,
                                     {function,
                                      #Fun<luerl_lib_basic.12.127600018>},
                                     {empty,<<"print">>,
                                      {function,
                                       #Fun<luerl_lib_basic.13.127600018>},
                                      empty}}},
                                   <<"rawequal">>,
                                   {function,
                                    #Fun<luerl_lib_basic.14.127600018>},
                                   {{{empty,<<"rawget">>,
                                      {function,
                                       #Fun<luerl_lib_basic.15.127600018>},
                                      empty},
                                     <<"rawlen">>,
                                     {function,
                                      #Fun<luerl_lib_basic.16.127600018>},
                                     {empty,<<"rawset">>,
                                      {function,
                                       #Fun<luerl_lib_basic.17.127600018>},
                                      empty},
                                     <<"register_hook">>,
                                     {function,0,0,[],[],
                                      [{push_last_lit,
                                        <<"on_message_publish">>},
                                       {return,1}]},
                                     {empty,<<"require">>,
                                      {function,
                                       #Fun<luerl_lib_package.0.128526133>},
                                      empty}},
                                    <<"select">>,
                                    {function,
                                     #Fun<luerl_lib_basic.18.127600018>},
                                    {{empty,<<"setmetatable">>,
                                      {function,
                                       #Fun<luerl_lib_basic.19.127600018>},
                                      empty},
                                     <<"string">>,
                                     {tref,9},
                                     {empty,<<"table">>,{tref,11},empty}},
                                    <<"tonumber">>,
                                    {function,
                                     #Fun<luerl_lib_basic.20.127600018>},
                                    {{empty,<<"tostring">>,
                                      {function,
                                       #Fun<luerl_lib_basic.21.127600018>},
                                      empty},
                                     <<"type">>,
                                     {function,
                                      #Fun<luerl_lib_basic.22.127600018>},
                                     {empty,<<"unpack">>,
                                      {function,
                                       #Fun<luerl_lib_basic.23.127600018>},
                                      empty}}}},
                                  nil},
                                 {table,
                                  {array,3,10,nil,
                                   {nil,
                                    {function,
                                     #Fun<luerl_lib_package.2.128526133>},
                                    {function,
                                     #Fun<luerl_lib_package.3.128526133>},
                                    nil,nil,nil,nil,nil,nil,nil}},
                                  empty,nil},
                                 {table,
                                  {array,0,10,nil,10},
                                  {{{empty,<<"_G">>,{tref,0},empty},
                                    <<"bit32">>,
                                    {tref,5},
                                    {empty,<<"debug">>,{tref,12},empty},
                                    <<"io">>,
                                    {tref,6},
                                    {empty,<<"math">>,{tref,7},empty}},
                                   <<"os">>,
                                   {tref,8},
                                   {{empty,<<"package">>,{tref,4},empty},
                                    <<"string">>,
                                    {tref,9},
                                    {empty,<<"table">>,{tref,11},empty}}},
                                  nil},
                                 {table,{array,0,10,nil,10},empty,nil},
                                 {table,
                                  {array,0,10,nil,10},
                                  {{empty,<<"config">>,<<"/\n;\n?\n!\n-\n">>,
                                    empty},
                                   <<"loaded">>,
                                   {tref,2},
                                   {empty,<<"path">>,
                                    <<"./?.lua;./?/init.lua">>,empty},
                                   <<"preload">>,
                                   {tref,3},
                                   {empty,<<"searchers">>,
                                    {tref,1},
                                    empty,<<"searchpath">>,
                                    {function,
                                     #Fun<luerl_lib_package.1.128526133>},
                                    empty}},
                                  nil},
                                 {table,
                                  {array,0,10,nil,10},
                                  {{{empty,<<"arshift">>,
                                     {function,
                                      #Fun<luerl_lib_bit32.7.103630370>},
                                     empty,<<"band">>,
                                     {function,
                                      #Fun<luerl_lib_bit32.0.103630370>},
                                     empty},
                                    <<"bnot">>,
                                    {function,
                                     #Fun<luerl_lib_bit32.1.103630370>},
                                    {empty,<<"bor">>,
                                     {function,
                                      #Fun<luerl_lib_bit32.2.103630370>},
                                     empty}},
                                   <<"btest">>,
                                   {function,
                                    #Fun<luerl_lib_bit32.3.103630370>},
                                   {{empty,<<"bxor">>,
                                     {function,
                                      #Fun<luerl_lib_bit32.4.103630370>},
                                     empty},
                                    <<"extract">>,
                                    {function,
                                     #Fun<luerl_lib_bit32.10.103630370>},
                                    {empty,<<"lrotate">>,
                                     {function,
                                      #Fun<luerl_lib_bit32.8.103630370>},
                                     empty}},
                                   <<"lshift">>,
                                   {function,
                                    #Fun<luerl_lib_bit32.5.103630370>},
                                   {{empty,<<"replace">>,
                                     {function,
                                      #Fun<luerl_lib_bit32.11.103630370>},
                                     empty},
                                    <<"rrotate">>,
                                    {function,
                                     #Fun<luerl_lib_bit32.9.103630370>},
                                    {empty,<<"rshift">>,
                                     {function,
                                      #Fun<luerl_lib_bit32.6.103630370>},
                                     empty}}},
                                  nil},
                                 {table,
                                  {array,0,10,nil,10},
                                  {empty,<<"flush">>,
                                   {function,#Fun<luerl_lib_io.0.23971557>},
                                   empty,<<"write">>,
                                   {function,#Fun<luerl_lib_io.1.23971557>},
                                   empty},
                                  nil},
                                 {table,
                                  {array,0,10,nil,10},
                                  {{{{empty,<<"abs">>,
                                      {function,
                                       #Fun<luerl_lib_math.2.133847239>},
                                      empty},
                                     <<"acos">>,
                                     {function,
                                      #Fun<luerl_lib_math.3.133847239>},
                                     {empty,<<"asin">>,
                                      {function,
                                       #Fun<luerl_lib_math.4.133847239>},
                                      empty}},
                                    <<"atan">>,
                                    {function,
                                     #Fun<luerl_lib_math.5.133847239>},
                                    {{empty,<<"atan2">>,
                                      {function,
                                       #Fun<luerl_lib_math.6.133847239>},
                                      empty},
                                     <<"ceil">>,
                                     {function,
                                      #Fun<luerl_lib_math.7.133847239>},
                                     {empty,<<"cos">>,
                                      {function,
                                       #Fun<luerl_lib_math.8.133847239>},
                                      empty}}},
                                   <<"cosh">>,
                                   {function,#Fun<luerl_lib_math.9.133847239>},
                                   {{{empty,<<"deg">>,
                                      {function,
                                       #Fun<luerl_lib_math.10.133847239>},
                                      empty},
                                     <<"exp">>,
                                     {function,
                                      #Fun<luerl_lib_math.11.133847239>},
                                     {empty,<<"floor">>,
                                      {function,
                                       #Fun<luerl_lib_math.12.133847239>},
                                      empty}},
                                    <<"fmod">>,
                                    {function,
                                     #Fun<luerl_lib_math.13.133847239>},
                                    {{empty,<<"frexp">>,
                                      {function,
                                       #Fun<luerl_lib_math.14.133847239>},
                                      empty},
                                     <<"huge">>,1.7976931348623157e308,
                                     {empty,<<"ldexp">>,
                                      {function,
                                       #Fun<luerl_lib_math.15.133847239>},
                                      empty}}},
                                   <<"log">>,
                                   {function,
                                    #Fun<luerl_lib_math.16.133847239>},
                                   {{{empty,<<"log10">>,
                                      {function,
                                       #Fun<luerl_lib_math.17.133847239>},
                                      empty},
                                     <<"max">>,
                                     {function,
                                      #Fun<luerl_lib_math.0.133847239>},
                                     {empty,<<"min">>,
                                      {function,
                                       #Fun<luerl_lib_math.1.133847239>},
                                      empty}},
                                    <<"modf">>,
                                    {function,
                                     #Fun<luerl_lib_math.18.133847239>},
                                    {{empty,<<"pi">>,3.141592653589793,empty},
                                     <<"pow">>,
                                     {function,
                                      #Fun<luerl_lib_math.19.133847239>},
                                     {empty,<<"rad">>,
                                      {function,
                                       #Fun<luerl_lib_math.20.133847239>},
                                      empty}},
                                    <<"random">>,
                                    {function,
                                     #Fun<luerl_lib_math.21.133847239>},
                                    {{empty,<<"randomseed">>,
                                      {function,
                                       #Fun<luerl_lib_math.22.133847239>},
                                      empty},
                                     <<"sin">>,
                                     {function,
                                      #Fun<luerl_lib_math.23.133847239>},
                                     {empty,<<"sinh">>,
                                      {function,
                                       #Fun<luerl_lib_math.24.133847239>},
                                      empty},
                                     <<"sqrt">>,
                                     {function,
                                      #Fun<luerl_lib_math.25.133847239>},
                                     {empty,<<"tan">>,
                                      {function,
                                       #Fun<luerl_lib_math.26.133847239>},
                                      empty,<<"tanh">>,
                                      {function,
                                       #Fun<luerl_lib_math.27.133847239>},
                                      empty}}}},
                                  nil},
                                 {table,
                                  {array,0,10,nil,10},
                                  {{empty,<<"clock">>,
                                    {function,#Fun<luerl_lib_os.0.78142656>},
                                    empty},
                                   <<"date">>,
                                   {function,#Fun<luerl_lib_os.1.78142656>},
                                   {empty,<<"difftime">>,
                                    {function,#Fun<luerl_lib_os.2.78142656>},
                                    empty},
                                   <<"getenv">>,
                                   {function,#Fun<luerl_lib_os.3.78142656>},
                                   {empty,<<"time">>,
                                    {function,#Fun<luerl_lib_os.4.78142656>},
                                    empty}},
                                  nil},
                                 {table,
                                  {array,0,10,nil,10},
                                  {{{empty,<<"byte">>,
                                     {function,
                                      #Fun<luerl_lib_string.0.17436618>},
                                     empty},
                                    <<"char">>,
                                    {function,
                                     #Fun<luerl_lib_string.1.17436618>},
                                    {empty,<<"dump">>,
                                     {function,
                                      #Fun<luerl_lib_string.2.17436618>},
                                     empty}},
                                   <<"find">>,
                                   {function,
                                    #Fun<luerl_lib_string.3.17436618>},
                                   {{empty,<<"format">>,
                                     {function,
                                      #Fun<luerl_lib_string.4.17436618>},
                                     empty},
                                    <<"gmatch">>,
                                    {function,
                                     #Fun<luerl_lib_string.5.17436618>},
                                    {empty,<<"gsub">>,
                                     {function,
                                      #Fun<luerl_lib_string.6.17436618>},
                                     empty}},
                                   <<"len">>,
                                   {function,
                                    #Fun<luerl_lib_string.7.17436618>},
                                   {{empty,<<"lower">>,
                                     {function,
                                      #Fun<luerl_lib_string.8.17436618>},
                                     empty},
                                    <<"match">>,
                                    {function,
                                     #Fun<luerl_lib_string.9.17436618>},
                                    {empty,<<"rep">>,
                                     {function,
                                      #Fun<luerl_lib_string.10.17436618>},
                                     empty},
                                    <<"reverse">>,
                                    {function,
                                     #Fun<luerl_lib_string.11.17436618>},
                                    {empty,<<"sub">>,
                                     {function,
                                      #Fun<luerl_lib_string.12.17436618>},
                                     empty,<<"upper">>,
                                     {function,
                                      #Fun<luerl_lib_string.13.17436618>},
                                     empty}}},
                                  nil}},
                                {{table,
                                  {array,0,10,nil,10},
                                  {empty,<<"__index">>,{tref,9},empty},
                                  nil},
                                 {table,
                                  {array,0,10,nil,10},
                                  {{empty,<<"concat">>,
                                    {function,
                                     #Fun<luerl_lib_table.0.108233019>},
                                    empty},
                                   <<"insert">>,
                                   {function,
                                    #Fun<luerl_lib_table.1.108233019>},
                                   {empty,<<"pack">>,
                                    {function,
                                     #Fun<luerl_lib_table.2.108233019>},
                                    empty},
                                   <<"remove">>,
                                   {function,
                                    #Fun<luerl_lib_table.3.108233019>},
                                   {empty,<<"sort">>,
                                    {function,
                                     #Fun<luerl_lib_table.4.108233019>},
                                    empty,<<"unpack">>,
                                    {function,
                                     #Fun<luerl_lib_table.5.108233019>},
                                    empty}},
                                  nil},
                                 {table,
                                  {array,0,10,nil,10},
                                  {{empty,<<"getmetatable">>,
                                    {function,
                                     #Fun<luerl_lib_debug.0.81249568>},
                                    empty},
                                   <<"getuservalue">>,
                                   {function,#Fun<luerl_lib_debug.1.81249568>},
                                   {empty,<<"setmetatable">>,
                                    {function,
                                     #Fun<luerl_lib_debug.2.81249568>},
                                    empty,<<"setuservalue">>,
                                    {function,
                                     #Fun<luerl_lib_debug.3.81249568>},
                                    empty}},
                                  nil},
                                 undefined,undefined,undefined,undefined,
                                 undefined,undefined,undefined},
                                10,10,10,10,10,10,10,10,10}},
                              [],13,
                              {array,0,10,undefined,10},
                              [],0,
                              {tref,0},
                              [],
                              {meta,nil,nil,nil,{tref,10}},
                              #Ref<10661.4274420810.282329091.250659>}]],
                           [{file,"src/emqx_hooks.erl"},{line,64}]},
                          {emqx_lua_hook,do_register_hooks,3,
                           [{file,"src/emqx_lua_hook.erl"},{line,162}]},
                          {emqx_lua_hook,do_load,1,
                           [{file,"src/emqx_lua_hook.erl"},{line,126}]},
                          {emqx_lua_hook,'-do_loadall/0-lc$^0/1-0-',1,
                           [{file,"src/emqx_lua_hook.erl"},{line,111}]},
                          {emqx_lua_hook,do_loadall,0,
                           [{file,"src/emqx_lua_hook.erl"},{line,111}]},
                          {emqx_lua_hook,handle_call,3,
                           [{file,"src/emqx_lua_hook.erl"},{line,61}]},
                          {gen_server,try_handle_call,4,
                           [{file,"gen_server.erl"},{line,661}]},
                          {gen_server,handle_msg,6,
                           [{file,"gen_server.erl"},{line,690}]}]},
                        {gen_server,call,[emqx_lua_hook,load_scripts]}}}}}}
```

**Comment by jdavidagudelo** [comment]
Yes it is emqx3.0, in emqx2.0 the LUA hooks work fine. I used the command ./bin/emqx_ctl plugins load emqx_lua_hook in emqx3.0 and it throws an error.

**Comment by tigercl** [comment]
Hi, @jdavidagudelo, we have fixed the bug of loading lua scripts in [3.0-rc.3](http://emqtt.com/downloads)
